src/vm/object.h

```diff
@@ -2127,17 +2127,14 @@ typedef CriticalHandle * CRITICALHANDLEREF;
 // Base class for WaitHandle 
 class WaitHandleBase :public MarshalByRefObjectBaseObject
 {
-    friend class WaitHandleNative;
     friend class MscorlibBinder;
 
 public:
-    __inline LPVOID GetWaitHandle() {LIMITED_METHOD_CONTRACT; return m_handle;}
+    __inline LPVOID GetWaitHandle() {LIMITED_METHOD_CONTRACT; return m_safeHandle != NULL ? m_safeHandle->GetHandle() : INVALID_HANDLE_VALUE;}
```

Comment:
This can crash the runtime on race conditions. This may need to use volatile read.

We are ok with null reference exceptions on race conditions in C#, but unmanaged side of the runtime should aoid accessing null pointers in all cases. Accessing null pointer in the unmanaged runtime will produce failfast that is assumed to be GC bug or data corruption - hard to people to diagnose on their own.

Reply:
I will fix that.

Reply:
Looked easier than it is done. I cannot use use `Volatile<SAFEHANDLEREF> m_safeHandle` because it breaks on checked builds where `SAFEHANDLEREF` is declared as `REF<SafeHandle>` and not as a simple pointer. Using `Volatile<SAFEHANDLE*> m_safeHandle` will remove the additional checked build checks, which is not ideal. Finally, the naïve version of keeping `SAFEHANDLEREF` and doing this:

```c
    __inline LPVOID GetWaitHandle() {
        LIMITED_METHOD_CONTRACT;
        SAFEHANDLEREF safeHandle = m_safeHandle;
        return safeHandle != NULL ? safeHandle->GetHandle() : INVALID_HANDLE_VALUE;
    }
```

is basically sure to run into compilers doing improper optimizations (https://github.com/dotnet/coreclr/pull/16089). So, any preferred way to fix that?

Reply:
Something like this?

```c++
class WaitHandleBase :public MarshalByRefObjectBaseObject
{
    friend class MscorlibBinder;

public:
    __inline LPVOID GetWaitHandle() {
        LIMITED_METHOD_CONTRACT;
        SAFEHANDLEREF safeHandle = (SAFEHANDLEREF)m_safeHandle.LoadWithoutBarrier();
        return safeHandle != NULL ? safeHandle->GetHandle() : INVALID_HANDLE_VALUE;
    }
    __inline SAFEHANDLEREF GetSafeHandle() {LIMITED_METHOD_CONTRACT; return (SAFEHANDLEREF)m_safeHandle.LoadWithoutBarrier();}

private:
    Volatile<SafeHandle*> m_safeHandle;
};
```